PG-1411 Make pg_resetwal work with TDE

artemgavrilov opened the pull request:
https://perconadev.atlassian.net/browse/PG-1411

The only thing that differs in pg_resetwal for TDE setups is that it performs WAL key rotation the same way as server itself on startup. During this rotation pg_resetwal always create `unecrypted` type of key as WAL will contain only one record that don't contain anything sensitive. So no need actually encrypt anything yet. On startup server will create new key with proper type.

dAdAbird commented on contrib/pg_tde/src/access/pg_tde_xlog_smgr.c:
```diff
@@ -240,6 +240,26 @@ tdeheap_xlog_seg_write(int fd, const void *buf, size_t count, off_t offset,
 	if (EncryptXLog)
 		return TDEXLogWriteEncryptedPages(fd, buf, count, offset, tli, segno);
 	else
+#else
+	InternalKey *key = pg_tde_read_last_wal_key();
+
+	if (key && key->type == TDE_KEY_TYPE_WAL_ENCRYPTED)
+	{
+		TDEPrincipalKey *principal_key;
+		InternalKey encryptionKey;
+
+		char		iv_prefix[16];
+		char		enc_buff[XLOG_BLCKSZ];
+		void	   *crypt_ctx = NULL;
+
+		CalcXLogPageIVPrefix(tli, segno, key->base_iv, iv_prefix);
+		pg_tde_stream_crypt(iv_prefix, offset,
+							(char *) buf, count,
+							enc_buff, key, &crypt_ctx);
+
+		return pg_pwrite(fd, enc_buff, count, offset);
+	}
+	else
```
I took a bit different approach in my [pg_basebackup PR](https://github.com/percona/postgres/pull/473/commits/1a52ab4c49920961d137eebfafe754fdb71292f8#diff-2c48cf024ce30498f9a9b821e900484e01de35f5ddbff82d0564559a3a68561eR1134). I created a helper function that de/crypts a given buffer instead of changing the write func. It's a kinda necessity in pg_basbackup case because it can have different write methods (w/ into tar, gzipped tar, lz4, etc), and it's easier to encrypt a buffer than redo all interfaces. 
In your case, it would eliminate the need in 
```

 #ifdef PERCONA_EXT
 	/* If we used xlog smgr, we need to update the file offset */
 	lseek(fd, XLOG_BLCKSZ, SEEK_CUR);
 #endif
```
which isn't that much of a win, but still one `#ifdef` less.

I don't say that my approach is better. But it definitely makes sense to unify approaches

artemgavrilov replied:
Changed in favor of latest merged changes that exposed required functions to FE.

artemgavrilov committed:
PG-1411 Make pg_resetwal work with TDE

As pg_resetwal removes old WAL segments and creates new one with empty
record we can do that write in unencrypted mode. However that requires
new WAL key creation in case if encryption was enabled before.

artemgavrilov committed:
PG-1411 Add pg_resetwal tap tests for TDE setup

These tests are copy of original pg_resetwal tests with enalbed WAL
encryption and removed flags validation as we interested here only in
proper enrypted WAL handling.

dutow commented on src/bin/pg_resetwal/Makefile:
```diff
@@ -15,14 +15,29 @@ subdir = src/bin/pg_resetwal
 
 OBJS = \
 	$(WIN32RES) \
+	xlogreader.o \
```
shouldn't this be also part of the if percona ext part later?

artemgavrilov replied:
Yep, fixed.

dAdAbird commented on src/bin/pg_resetwal/Makefile:
```diff
@@ -15,14 +15,29 @@ subdir = src/bin/pg_resetwal
 top_builddir = ../../..
 include $(top_builddir)/src/Makefile.global
 
+override CPPFLAGS := -DFRONTEND $(CPPFLAGS)
```
I think it would be moved into `ifeq ($(enable_percona_ext),yes)` clause as -DFRONTEND needed here only for `xlogreader`

artemgavrilov replied:
Yep, moved.

dAdAbird commented on src/bin/pg_resetwal/pg_resetwal.c:
```diff
@@ -488,6 +504,15 @@ main(int argc, char *argv[])
 
+#ifdef PERCONA_EXT
+	/*
+	 * As we always write empty record here we can do it in unencrypted mode.
```
We write not empty but checkpoint record here, innit?

artemgavrilov replied:
Yes, but for some reason it's how this named in pg_rewind sources:
```
/*
 * Write an empty XLOG file, containing only the checkpoint record
 * already set up in ControlFile.
 */
static void
WriteEmptyXLOG(void)
```

Anyway changed this comment to be more clear

dAdAbird commented on src/bin/pg_resetwal/pg_resetwal.c:
```diff
@@ -488,6 +504,15 @@ main(int argc, char *argv[])
+#ifdef PERCONA_EXT
+	/*
+	 * As we always write empty record here we can do it in unencrypted mode.
+	 * We initialize writes here because new WAL key may be created in some cases,
+	 * and we want it only when we are ready to perform acutal writes.
```
I struggle to understand this comment tbh. I guess, although we won't encrypt anything, we still need a new key and to write via xlog_smgr->seg_write so when we start the server, TDE can identify that this segment doesn't need decryption.

What do you mean by "new WAL key may be created in some cases,  and we want it only when we are ready to perform acutal writes."?

artemgavrilov replied:
Terrible sentence for sure. I wanted to point why writes initialized on this line and not together with `TDEXLogSmgrInit` above. We want `TDEXLogSmgrInitWrite` to be invoked only when we checked all arguments and other stuff because this function will create new key **if** encryption was enabled before. If it wasn't no new key will be added.

Rewrote this section.

dAdAbird commented on src/bin/pg_resetwal/pg_resetwal.c:
```diff
@@ -488,6 +504,18 @@ main(int argc, char *argv[])
+#ifdef PERCONA_EXT
+	/*
+	 * The only thing that WAL will contain after reset is a checkpoint, so we can
+	 * write it in unencrypted form. There are no sensitive data in it.
```
"There **_is_** no sensitive data..."

artemgavrilov replied:
Fixed

dAdAbird commented on src/bin/pg_resetwal/pg_resetwal.c:
```diff
@@ -488,6 +504,18 @@ main(int argc, char *argv[])
+	 * The only thing that WAL will contain after reset is a checkpoint, so we can
+	 * write it in unencrypted form. There are no sensitive data in it.
+	 *
+	 * We doing writes initialization only here and not at the startup because we 
```
"We **_are_** doing **_a write_** initialization..."

artemgavrilov replied:
Fixed

dAdAbird commented on src/bin/pg_resetwal/pg_resetwal.c:
```diff
@@ -488,6 +504,18 @@ main(int argc, char *argv[])
+	 *
+	 * We doing writes initialization only here and not at the startup because we 
+	 * want to be sure that everything is checked and ready for writing at this point.
+	 * This function will create new WAL key if WAL encryption is enabled.
```
I'd also add the explanation why do we need all these TDEXLogSmgrInitWrite and xlog_smgr->seg_write if we do unencrypted write. It's not that obvious as it may seen to us. Tbh, my first reaction was "why all of these for an unencrypted write?" but then "ahh, yes, the past and future encrypted WAL..."

artemgavrilov replied:
Added the explanation

dAdAbird reviewed (commented):
a bit more comment nitpicking)
